sui_core/src/consensus_adapter.rs

```diff
@@ -129,91 +129,91 @@ impl ConsensusAdapter {
         }
 
         // Wait for the consensus to sequence the certificate and assign locks to shared objects.
-        timeout(self.max_delay, receiver)
-            .await
-            .map_err(|e| SuiError::ConsensusConnectionBroken(e.to_string()))?
-            .expect("Channel with consensus listener dropped")
+        match timeout(self.max_delay, receiver).await {
+            Ok(reply) => reply.expect("Channel with consensus listener dropped"),
+            Err(e) => {
+                let message = ConsensusListenerMessage::Cleanup(serialized);
+                self.tx_consensus_listener
+                    .send(message)
+                    .await
+                    .expect("Channel with consensus listener dropped");
+                Err(SuiError::ConsensusConnectionBroken(e.to_string()))
+            }
+        }
     }
 }
 
 /// This module interfaces the consensus with Sui. It receives certificates input to consensus and
 /// notify the called when they are sequenced.
 pub struct ConsensusListener {
-    /// The authority state to execute shared-object transactions.
-    state: Arc<AuthorityState>,
     /// Receive messages input to the consensus.
-    rx_consensus_input: Receiver<ConsensusInput>,
+    rx_consensus_input: Receiver<ConsensusListenerMessage>,
     /// Receive consensus outputs.
     rx_consensus_output: Receiver<ConsensusOutput>,
+    /// The maximum number of pending replies.
+    max_pending_transactions: usize,
     /// Keep a map of all consensus inputs that are currently being sequenced.
     pending: HashMap<ConsensusTransactionDigest, Vec<Replier>>,
```

Comment:
Should we keep the full certificate instead of just the digest? We might want to re-submit, or share at some point? Should this be persisted to disk, or are we ok if it goes away upon crash.

Reply:
It will be the job of the client (and gateway) to handle resubmit in case of timeouts or crashes. It is ok if this goes away upon crash, the client will simply have to re-submit the certificate.